src/test/spec/unified_runner/observer.rs

```diff
@@ -0,0 +1,103 @@
+use tokio::sync::{
+    broadcast::{
+        self,
+        error::{RecvError, TryRecvError},
+    },
+    RwLock,
+};
+
+use std::{sync::Arc, time::Duration};
+
+use crate::{
+    error::{Error, Result},
+    runtime,
+    test::Event,
+};
+
+use super::{events_match, EntityMap, ExpectedEvent};
+
+/// Observer used to cache all the seen events for a given client in a unified test.
+/// Used to implement assertEventCount and waitForEvent operations.
+#[derive(Debug)]
+pub(crate) struct EventObserver {
+    seen_events: Vec<Event>,
+    receiver: broadcast::Receiver<Event>,
+}
+
+impl EventObserver {
+    pub fn new(receiver: broadcast::Receiver<Event>) -> Self {
+        Self {
+            seen_events: Vec::new(),
+            receiver,
+        }
+    }
+
+    pub(crate) async fn recv(&mut self) -> Option<Event> {
+        match self.receiver.recv().await {
+            Ok(e) => {
+                self.seen_events.push(e.clone());
+                Some(e)
+            }
+            Err(RecvError::Lagged(_)) => panic!("event receiver lagged"),
+            Err(RecvError::Closed) => None,
+        }
+    }
+
+    fn try_recv(&mut self) -> Option<Event> {
+        match self.receiver.try_recv() {
+            Ok(e) => {
+                self.seen_events.push(e.clone());
+                Some(e)
+            }
+            Err(TryRecvError::Lagged(_)) => panic!("event receiver lagged"),
+            Err(TryRecvError::Closed | TryRecvError::Empty) => None,
+        }
+    }
+
+    pub(crate) async fn event_count(
```

Comment:
nit: can we make the names here and for `wait_for_event` a little more descriptive like `matching_event_count` and `wait_for_matching_events` or something?